aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp

```diff
@@ -0,0 +1,60 @@
+#include <ATen/ATen.h>
+#include <ATen/core/Type.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/cpp_custom_type_hack.h>
+#include <ATen/native/quantized/cpu/fbgemm_utils.h>
+#include <ATen/quantized/Quantizer.h>
+
+#include <algorithm>
+#include <vector>
+
+namespace at {
+namespace native {
+namespace {
+
+class QFCUnpackWeightInt8 final : public c10::OperatorKernel {
+ public:
+#ifdef USE_FBGEMM
+  at::Tensor operator()(at::Tensor packed_weight) {
+    // Pull out the PackBMatrix instance from the owning tensor.
+    auto& pack_ptr = cpp_custom_type_hack::cast<PackedFCWeight>(packed_weight);
+    auto packB = pack_ptr.w.get();
+
+    int64_t N = static_cast<int64_t>(packB->numCols());
+    int64_t K = static_cast<int64_t>(packB->numRows());
+
+    float weight_scale_float = pack_ptr.w_scale;
+    int32_t weight_zero_point_int32 = pack_ptr.w_zp;
+
+    auto weight_origin = _empty_affine_quantized(
+        {N, K},
+        at::device(kCPU).dtype(kQInt8),
+        weight_scale_float,
+        weight_zero_point_int32);
+    int8_t* weight_ptr_int8 =
+        reinterpret_cast<int8_t*>(weight_origin.data<c10::qint8>());
+
+    packB->unpack(weight_ptr_int8);
```

Comment:
do you have to unpack weight if we store weight in int8?

Reply:
Yes: pack means the memory layout has been changed. We have to recover the original memory layout from the packed buffers.